### PR TITLE
Additions to the Perl family of languages

### PR DESCRIPTION
**Pull request by JJ**
Adding some extensions and sample files for Perl

## Description

- Moves Rexfile to the correct language
- Adds extensions to Perl and Perl6. Some sample files, too.

## Checklist:

- [x] **I am associating a language with a new file extension.** (but search for pod6 does not seem to work very well)
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/perl6/doc/blob/master/doc/Type/Any.pod6
    - Sample license(s): ARTISTIC https://github.com/perl6/doc/blob/master/LICENSE
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -3369,6 +3369,8 @@ Perl:
   - ".psgi"
   - ".t"
   filenames:
+  - Makefile.PL
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: This is redundant, because it'll be matched against the `*.pl` file extension.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -6,11 +6,15 @@ Contributions to this project are [released](https://help.github.com/articles/gi
 
 This project adheres to the [Contributor Covenant Code of Conduct](http://contributor-covenant.org/). By participating, you are expected to uphold this code.
 
-The majority of contributions won't need to touch any Ruby code at all.
```
Comment (Alhadis) [inline CONTRIBUTING.md]: I see you first forked Linguist in [November 2015](https://github.com/github/linguist/pull/4066/commits) (!!). You're overwriting unrelated changes that have been made to the docs since then. Please update `CONTRIBUTING.md` with the [latest version](https://github.com/github/linguist/blob/4f0f9bd51d9bf54e05d1b1456dfdc3df54b3360d/CONTRIBUTING.md).

File: lib/linguist/languages.yml [hunk 3]
```diff
@@ -3369,6 +3369,8 @@ Perl:
...
+  - Rexfile
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: In-the-wild usage is a little questionable... only [~155 results](https://github.com/search?q=filename%3ARexfile+NOT+nothack&type=Code), although they appear to be well-distributed across users/repositories (which technically meets the criteria of "being used in hundreds of repositories"). I think it's okay, particularly since it's an uncommon filename unlikely to cause conflict. @pchaigno, thoughts?

File: lib/linguist/languages.yml [hunk 4]
```diff
@@ -3387,9 +3389,8 @@ Perl 6:
   - ".pl6"
   - ".pm"
   - ".pm6"
+  - ".pod6"
```
Comment (Alhadis) [inline lib/linguist/languages.yml]: This was discussed before, and should be omitted. See #3366.

File: lib/linguist/languages.yml (hunk 1 above)
Comment (pchaigno) [reply lib/linguist/languages.yml]: @Alhadis That extension is shared between Perl, Perl 6, and Prolog. Adding the `Makefile.PL` might improve detection for that particular case. Are we sure `Makefile.PL` is always Perl and not Perl 6 though?

File: CONTRIBUTING.md (hunk 2 above)
Comment (pchaigno) [reply CONTRIBUTING.md]: @JJ You might want to update your fork...

File: lib/linguist/languages.yml (hunk 3 above)
Comment (pchaigno) [reply lib/linguist/languages.yml]: I'll defer to @lildude :smile:

File: CONTRIBUTING.md (hunk 2 above)
Comment (JJ) [reply CONTRIBUTING.md]: I'll do that :+1:

File: lib/linguist/languages.yml (hunk 1 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: I stand corrected, then. :wink: @pchaigno knows more about the Bayesian classifier than I do.

I don't recall ever seeing a Perl 6 version of `Makefile.PL`... it's always ever been used by Perl 5 modules downloaded/installed from CPAN. I could be wrong though.

Comment (JJ) [reply lib/linguist/languages.yml]: Pretty sure. Perl6 has pretty much normalized to .pl6 or .p6 now. It's not using this particular file, either. This is the repo for all modules https://github.com/moritz/perl6-all-modules/search?utf8=%E2%9C%93&q=%22Makefile.PL%22&type= and it's not there even once. Makefile.PL is kind of a Rakefile used for some CPAN distributions; Perl6 uses zef and META6.json

File: lib/linguist/languages.yml (hunk 4 above)
Comment (JJ) [reply lib/linguist/languages.yml]: Checking it out... And I would kindly ask you to reconsider. What brought me (back) here was github/markup#1173, where I spent quite a long time until I discovered .pod6 files were not recognized as Perl6, which had me do a workaround... In fact, I fear .pod6 files will not be recognized as such, even by markup, if that extension is not included here. pod6 is used extensively in https://github.com/perl6/doc, if nowhere else, and as indicated in #3366, it would help adoption of the language. I can wait for that one to be accepted, though... Or just eliminate that change and defer to that one.

File: lib/linguist/languages.yml (hunk 1 above)
Comment (pchaigno) [reply lib/linguist/languages.yml]: > @pchaigno knows more about the Bayesian classifier than I do.

I don't know much about the classifier either, but in this particular case, there's a [specific filename strategy](https://github.com/github/linguist/blob/dd3b1eec911072ca051ecccabc03a47bafa2c76c/lib/linguist.rb#L62) before we reach the classifier.

PS: Not sure why I keep using "particular case" today...

File: lib/linguist/languages.yml (hunk 3 above)
Comment (JJ) [reply lib/linguist/languages.yml]: Problem here is that Rexfile was actually in the Perl6 field (my fault: https://github.com/github/linguist/pull/2005/files). It should better be here, Rexfile is in no way Perl6 code.

File: lib/linguist/languages.yml (hunk 1 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: Sounds good to me, then. :+1: The bulk of my familiarity with `Makefile.PL` is limited to seeing those files whoosh past the terminal when installing CPAN modules... :grinning:

File: CONTRIBUTING.md (hunk 2 above)
Comment (JJ) [reply CONTRIBUTING.md]: I've restored the HEAD copy. In fact, advice to install CMAKE, which was the only actual change, is already in the README.md file.

File: lib/linguist/languages.yml (hunk 1 above)
Comment (JJ) [reply lib/linguist/languages.yml]: Is there any other change you want me to do here?

Comment (JJ) [reply lib/linguist/languages.yml]: Ping... Is there anything missing?

File: lib/linguist/languages.yml (hunk 4 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: Personally, I'm okay with adding `.pod6` as an extension. But since there's been no official consensus on the matter, it might be easier to omit this particular addition and defer to #3366 instead. =)

File: lib/linguist/languages.yml (hunk 1 above)
Comment (Alhadis) [reply lib/linguist/languages.yml]: No, it looks fine to me. For the sake of quicker review, it might be better to omit `.pod6` since #3366 is more appropriate for discussion its inclusion.

Comment (JJ) [reply lib/linguist/languages.yml]: Eliminated `.pod6`. I can try and work on that one, although I can't add or modify that pull request...

Comment (Alhadis) [reply lib/linguist/languages.yml]: Don't worry about #3366. :) I believe that PR needs formal approval from GitHub staff more than anything.

**Comment by JJ** [comment]
Thanks!

(But let me worry about that one, it's really important for the community)

File: lib/linguist/languages.yml (hunk 3 above)
Comment (lildude) [reply lib/linguist/languages.yml]: This is fine. It slipped under the radar before when it was initially added so no point removing it now.

File: samples/Perl 6/Any.pod6 [hunk 5]
```diff
@@ -0,0 +1,987 @@
+=begin pod
```
Comment (lildude) [inline samples/Perl 6/Any.pod6]: This file needs to be removed too as you've removed `.pod6` from this PR.

**Comment by JJ** [comment]
OK.

File: samples/Perl 6/Any.pod6 (hunk 5 above)
Comment (JJ) [reply samples/Perl 6/Any.pod6]: Done.

**Comment by lildude** [comment]
Pulling in the latest master into your fork will address the CI failure.  You'll also need to address the merge conflict.

**Comment by JJ** [comment]
Will do that. Thanks.

**Comment by JJ** [comment]
Please check the latest change that addresses conflict.

**Comment by JJ** [comment]
Fixes CI error about alphabetical order on the files.

**Comment by JJ** [comment]
Sorry, what else is left to do here?

**Comment by Alhadis** [comment]
> Sorry, what else is left to do here?

Wait for this long-ass harvest to finish over my primitive, non-NBN-enabled Australian Internet connection, and be patient.

As an alternative to patience, you could entertain yourself by racing it against the following Perl script and seeing which one finishes in your lifetime:

~~~perl
$_ = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
/a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*a*[nope]/;
~~~

**Comment by JJ** [comment]
I can do that in Perl 6, where infinity is just another data structure... Or I can wait for the [latest CI to finish with flying colors](https://travis-ci.org/github/linguist/builds/364160795?utm_source=github_status&utm_medium=notification), which it did! :penguin:

**Comment by Alhadis** [comment]
Better idea! Let's all figure out why negative zero exists, and what sick human being thought a zero needed a sign-bit, and wonder why the hell -0 is different to 0 in any fathomable universe.

We can't divide by zero, but we can pointlessly sign it as +0 or -0 instead. Mmmm'kay, IEEE 754, whatever you reckon...

I don't know why I'm flagellating myself thinking about the real-world difference of positive and negative zero when I should be [making wisecracks in commit logs](https://github.com/Alhadis/language-sed/commit/cfbb39480e6577dc46cbaf48399dedd142ac3d8c) like last night.

File: samples/Perl 6/chromosome.pl [hunk 6]
```diff
@@ -0,0 +1,9 @@
+class Chromosome {
+  has Seq $.chromosome is rw;
+  has $.fitness is rw;
+
+}
+
+my $len = 32;
+my $this-chromosome = Chromosome.new( chromosome => map( { rand >= 0.5 ?? True !! False }, 1..$len )  );
+say $this-chromosome.chromosome();
```
Comment (lildude) [inline samples/Perl 6/chromosome.pl]: Is this file somehow uniquely Perl 6 such that it can clearly be differentiated from the exact same code written in Perl 5, or should this be added to the `samples/Perl/`?

The reason I ask is we use these samples to train the classifier and if this is identical to Perl 5, it's likely to add noise to the Perl 6 side of the classifier when it comes down to a Perl 5 vs Perl 6 comparison.

Comment (pchaigno) [reply samples/Perl 6/chromosome.pl]: It could also go into `tests/fixtures/Perl 6/` if we don't want it as a sample.

Comment (JJ) [reply samples/Perl 6/chromosome.pl]: On a superficial level, it could look similar to Perl. It's been a long time, but if I remember correctly I included it here precisely for that reason. It could go to fixtures alright. Just let me know what to do, and I'll do it straight away.

Comment (lildude) [reply samples/Perl 6/chromosome.pl]: Ok. If this is just to keep the tests passing, then yup, `tests/fixtures/Perl 6/` is definitely the best place for it.

Comment (JJ) [reply samples/Perl 6/chromosome.pl]: Done. I'm getting some local errors which apparently have nothing to do with moving stuff around. Let me know if there's something I can do about that.

**Comment by JJ** [comment]
Yay! You made my day. Thanks!
